contracts/FlashRollover.sol

```diff
@@ -0,0 +1,344 @@
+        address[] calldata assets,
+        uint256[] calldata amounts,
+        uint256[] calldata premiums,
+        OperationData memory opData
+    ) internal returns (bool) {
+        OperationContracts memory opContracts = _getContracts(opData.isLegacy);
+
+        // Get loan details
+        LoanLibrary.LoanData memory loanData = opContracts.loanCore.getLoan(opData.loanId);
+        require(loanData.borrowerNoteId != 0, "Cannot find note");
+
+        address borrower = opContracts.borrowerNote.ownerOf(loanData.borrowerNoteId);
+        address lender = opContracts.lenderNote.ownerOf(loanData.lenderNoteId);
+
+        // Do accounting to figure out amount each party needs to receive
+        (uint256 flashAmountDue, uint256 needFromBorrower, uint256 leftoverPrincipal) = _ensureFunds(
+            amounts[0],
+            premiums[0],
+            opContracts.feeController.getOriginationFee(),
+            opData.newLoanTerms.principal
+        );
+
+        if (needFromBorrower > 0) {
+            require(IERC20(assets[0]).balanceOf(borrower) >= needFromBorrower, "Borrower cannot pay");
+        }
+
+        _repayLoan(opContracts, loanData);
+        uint256 newLoanId = _initializeNewLoan(opContracts, borrower, lender, loanData.terms.collateralTokenId, opData);
+
+        if (leftoverPrincipal > 0) {
+            IERC20(assets[0]).transfer(borrower, leftoverPrincipal);
+        } else if (needFromBorrower > 0) {
+            IERC20(assets[0]).transferFrom(borrower, address(this), needFromBorrower);
+        }
+
+        // Approve all amounts for flash loan repayment
+        IERC20(assets[0]).approve(address(LENDING_POOL), flashAmountDue);
+
+        emit Rollover(lender, borrower, loanData.terms.collateralTokenId, newLoanId);
+
+        if (opData.isLegacy) {
+            emit Migration(address(opContracts.loanCore), address(opContracts.newLoanLoanCore), newLoanId);
+        }
+
+        return true;
+    }
+
+    function _ensureFunds(
+        uint256 amount,
+        uint256 premium,
+        uint256 originationFee,
+        uint256 newPrincipal
+    )
+        internal
+        pure
+        returns (
+            uint256 flashAmountDue,
+            uint256 needFromBorrower,
+            uint256 leftoverPrincipal
+        )
+    {
+        // Make sure new loan, minus pawn fees, can be repaid
+        flashAmountDue = amount + premium;
+        uint256 willReceive = newPrincipal - ((newPrincipal * originationFee) / 10_000);
+
+        if (flashAmountDue > willReceive) {
+            // Not enough - have borrower pay the difference
+            needFromBorrower = flashAmountDue - willReceive;
+        } else if (willReceive > flashAmountDue) {
+            // Too much - will send extra to borrower
+            leftoverPrincipal = willReceive - flashAmountDue;
+        }
+
+        // Either leftoverPrincipal or needFromBorrower should be 0
+        require(leftoverPrincipal & needFromBorrower == 0, "_ensureFunds computation");
```

Comment:
couldn't there be some unlikely circumstance where both are nonzero but their bitwise AND cancels out to 0? I would expend the extra like 20 gas to fully check

Reply:
Good catch and true - will fix